### PR TITLE
Scrutinizer: enable getTxByHash, store hash with TxReference

### PR DESCRIPTION
**Pull request by NateWilliams2**
Addresses #340

### CODE REVIEW

File: vochain/scrutinizer/transaction.go [hunk 1]
```diff
@@ -27,9 +28,20 @@ func (s *Scrutinizer) GetTxReference(height uint64) (*indexertypes.TxReference,
 	return txReference, nil
 }
 
+// GetTxReference fetches the txReference for the given tx hash
+func (s *Scrutinizer) GetTxHashReference(hash types.HexBytes) (*indexertypes.TxReference, error) {
+	txReference := &indexertypes.TxReference{}
+	err := s.db.FindOne(txReference, badgerhold.Where("Hash").Eq(hash))
```
Comment (p4u) [inline vochain/scrutinizer/transaction.go]: missing Index("Hash") ?

**Review by p4u — CHANGES_REQUESTED** [review]
Can you make a benchmark such as: https://github.com/vocdoni/vocdoni-node/blob/master/vochain/scrutinizer/db_benchmark_test.go

In order to see if, when the DB grows, we keep the query time constant.

**Comment by NateWilliams2** [comment]
Benchmarking:
The good news is that indexing the txs is not significantly different:
`master`:
```BenchmarkCheckTx/indexTx-4         	     366	  67358328 ns/op	18087009 B/op	  286999 allocs/op```

`getTxByHash`:
```BenchmarkCheckTx/indexTx-4         	     352	  65903017 ns/op	17040037 B/op	  284965 allocs/op```

the bad news is that fetching txs by hash is absolutely terrible. Here's benchmarks of fetching transactions from a relatively small pool, both directly from the tx index, and from the tx hash:

```
2021-11-17T13:38:08-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 1000 total) by index, took 242.679896ms
2021-11-17T13:38:17-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 1000 total) by hash, took 9.288797947s
2021-11-17T13:38:20-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 2000 total) by index, took 165.261123ms
2021-11-17T13:38:36-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 2000 total) by hash, took 16.72569996s
2021-11-17T13:38:38-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 3000 total) by index, took 187.115295ms
2021-11-17T13:38:58-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 3000 total) by hash, took 19.82848869s
2021-11-17T13:39:01-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 4000 total) by index, took 311.170259ms
2021-11-17T13:39:27-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 4000 total) by hash, took 26.171746103s
2021-11-17T13:39:29-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 5000 total) by index, took 168.56385ms
2021-11-17T13:39:53-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 5000 total) by hash, took 24.006937486s
2021-11-17T13:39:55-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 6000 total) by index, took 170.16323ms
2021-11-17T13:40:23-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 6000 total) by hash, took 28.0696469s
2021-11-17T13:40:25-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 7000 total) by index, took 160.863577ms
2021-11-17T13:40:58-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 7000 total) by hash, took 32.956502359s
BenchmarkFetchTx/fetchTx-4         	       7	24689754785 ns/op	4840272009 B/op	106015677 allocs/op
```

As you can see, while thee "index-fetched" operations remain relatively constant, fetching transactions by hash gets very slow, very quickly. 

Without indexing by hash, it was even worse:
```
2021-11-17T13:45:05-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 1000 total) by index, took 180.734072ms
2021-11-17T13:45:19-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 1000 total) by hash, took 13.93434078s
2021-11-17T13:45:21-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 2000 total) by index, took 164.36524ms
2021-11-17T13:46:15-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 2000 total) by hash, took 53.843692464s
2021-11-17T13:46:17-05:00	INFO	scrutinizer/db_benchmark_test.go:105	fetched 1000 transactions (out of 3000 total) by index, took 493.735952ms
2021-11-17T13:47:58-05:00	INFO	scrutinizer/db_benchmark_test.go:112	fetched 1000 transactions (out of 3000 total) by hash, took 1m40.554100489s
```

I think it's fair to say this method is not scalable, at least while using badgerhold. cc @p4u

**Comment by NateWilliams2** [comment]
A solution I could think of would be to add another database specifically for storing Hash->Index, where Hash is the key, to preserve constant-time queries for the hash. This seems overkill to me.

**Comment by p4u** [comment]
@mvdan did a workaround for querying indexes, but it had some constrains I don't remember now. Maybe we doing some of the queries wrong... That timing is too horrible, I don't believe it :/

Also @mvdan is working on a replacement for badgerhold, I think it'll be ready soon.

**Comment by p4u** [comment]
Does changing the Index position make some difference?
```
	err := s.db.FindOne(txReference, badgerhold.Where("Hash").Index("Hash").Eq(hash))
```

**Comment by NateWilliams2** [comment]
> Does changing the Index position make some difference?
> 
> ```
> 	err := s.db.FindOne(txReference, badgerhold.Where("Hash").Index("Hash").Eq(hash))
> ```

The index actually has to go at the end of a complete query. For this query, there's only one position it can be. 
Also, we're using a forked version of badgerhold: https://github.com/vocdoni/badgerhold/commits/master
It includes two innovations: constant-time writes for indexes, and faster queries by primary key (you may remember there was a crazy bug where using `badgerhold.Key` rather than specifying the actual key string took exponentially longer). Unfortunately these two improvements are already in place and don't help with this situation.

**Comment by p4u** [comment]
> > Does changing the Index position make some difference?
> > ```
> > 	err := s.db.FindOne(txReference, badgerhold.Where("Hash").Index("Hash").Eq(hash))
> > ```
> 
> The index actually has to go at the end of a complete query. For this query, there's only one position it can be. Also, we're using a forked version of badgerhold: https://github.com/vocdoni/badgerhold/commits/master It includes two innovations: constant-time writes for indexes, and faster queries by primary key (you may remember there was a crazy bug where using `badgerhold.Key` rather than specifying the actual key string took exponentially longer). Unfortunately these two improvements are already in place and don't help with this situation.

Umm, in most of our Queries we don't have the Index() at the end: https://github.com/vocdoni/vocdoni-node/blob/master/vochain/scrutinizer/process.go#L107

**Comment by NateWilliams2** [comment]
`Badgerhold.Where()` returns a `badgerhold.Criterion`, which has the method `.Eq()` that returns a `badgerhold.Query`. The `badgerhold.Query` is the object that has the `.Index()` method. So for the examples you sent, we have many `badgerhold.Query`s strung together, so the `Index()` can go after any one of them. In this situation, though, there's only one query- the `Eq()`, so this is the only place the `Index()` could go.

**Comment by p4u** [comment]
So let's go ahead with this, event if it is very slow we should fix that soon with the badgerhold removal (in favor of sqlite).

@NateWilliams2 we still need to figure out a way to return the txHash to the user on "submitRawTx" method. Probably the best way is to make CheckTx() return the hash of the transaction (if accepted by the local mempool node) cause we already have the hash computed somewhere in the code (for adding it to the vote cache).

**Comment by p4u** [comment]
This branch should be rebased with the current master branch. There are some minor conflicts.

**Pull request merged**